File: Difficulty/A. Easy/0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
class Solution:
    def mergeTwoLists(self, list1: ListNode, list2: ListNode) -> ListNode:
        """
        Merge two sorted linked lists into one sorted linked list.

        Parameters:
        list1 (ListNode): The head of the first sorted linked list.
        list2 (ListNode): The head of the second sorted linked list.

        Returns:
        ListNode: The head of the merged sorted linked list.
        """
        # Initialize a dummy node to simplify edge cases
        dummy = ListNode()
        tail = dummy
        
        # Iterate over both lists until one is empty
        while list1 and list2:
            if list1.val <= list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next
        
        # Append any remaining nodes from either list
        if list1:
            tail.next = list1
        else:
            tail.next = list2
        
        return dummy.next

File: Explore Card Track/3. Linked List/0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

/**
 * @brief Merge two sorted linked lists into one sorted list.
 *
 * @param list1 (struct ListNode*): Pointer to the first sorted linked list.
 * @param list2 (struct ListNode*): Pointer to the second sorted linked list.
 *
 * @return struct ListNode*: Pointer to the merged sorted linked list.
 */
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    // Create a dummy node to act as the head of the merged list
    struct ListNode dummy;
    struct ListNode* tail = &dummy;
    dummy.next = NULL;

    // Traverse both lists and merge them
    while (list1 != NULL && list2 != NULL) {
        if (list1->val <= list2->val) {
            tail->next = list1;
            list1 = list1->next;
        } else {
            tail->next = list2;
            list2 = list2->next;
        }
        tail = tail->next;
    }

    // Append the remaining nodes from either list
    if (list1 != NULL) {
        tail->next = list1;
    } else {
        tail->next = list2;
    }

    return dummy.next;
}